Sources/WhisperKit/Core/SegmentSeeker.swift

```diff
@@ -149,9 +180,328 @@ public class SegmentSeeker: SegmentSeeking {
+        }
+
+        // Initialize cost matrix and trace matrix
+        var costMatrix = Array(repeating: Array(repeating: Double.infinity, count: numberOfColumns + 1), count: numberOfRows + 1)
+        var traceMatrix = Array(repeating: Array(repeating: -1, count: numberOfColumns + 1), count: numberOfRows + 1)
+
+        costMatrix[0][0] = 0
+        for i in 1...numberOfColumns {
+            traceMatrix[0][i] = 2
+        }
+        for i in 1...numberOfRows {
+            traceMatrix[i][0] = 1
+        }
+
+        for row in 1...numberOfRows {
+            for column in 1...numberOfColumns {
+                let matrixValue = -matrix[(row - 1) * numberOfColumns + (column - 1)].doubleValue
+                let costDiagonal = costMatrix[row - 1][column - 1]
+                let costUp = costMatrix[row - 1][column]
+                let costLeft = costMatrix[row][column - 1]
+
+                let (computedCost, traceValue) = minCostAndTrace(
+                    costDiagonal: costDiagonal,
+                    costUp: costUp,
+                    costLeft: costLeft,
+                    matrixValue: matrixValue
+                )
+
+                costMatrix[row][column] = computedCost
+                traceMatrix[row][column] = traceValue
+            }
+        }
+
+        let dtw = backtrace(fromTraceMatrix: traceMatrix)
+
+        return dtw
+    }
+
+    func minCostAndTrace(costDiagonal: Double, costUp: Double, costLeft: Double, matrixValue: Double) -> (Double, Int) {
+        let c0 = costDiagonal + matrixValue
+        let c1 = costUp + matrixValue
+        let c2 = costLeft + matrixValue
+
+        if c0 < c1 && c0 < c2 {
+            return (c0, 0)
+        } else if c1 < c0 && c1 < c2 {
+            return (c1, 1)
+        } else {
+            return (c2, 2)
+        }
+    }
+
+    func backtrace(fromTraceMatrix traceMatrix: [[Int]]) -> (textIndices: [Int], timeIndices: [Int]) {
+        var i = traceMatrix.count - 1
+        var j = traceMatrix[0].count - 1
+
+        var textIndices = [Int]()
+        var timeIndices = [Int]()
+
+        while i > 0 || j > 0 {
+            textIndices.append(i - 1)
+            timeIndices.append(j - 1)
+
+            switch traceMatrix[i][j] {
+                case 0:
+                    i -= 1
+                    j -= 1
+                case 1:
+                    i -= 1
+                case 2:
+                    j -= 1
+                default:
+                    break
+            }
+        }
+
+        return (textIndices.reversed(), timeIndices.reversed())
+    }
+
+    func mergePunctuations(alignment: [WordTiming], prepended: String, appended: String) -> [WordTiming] {
+        var mergedAlignment = [WordTiming]()
+
+        // Merge prepended punctuations
+        for i in 1..<alignment.count {
+            let currentWord = alignment[i]
+            if i > 0, currentWord.word.starts(with: " "), prepended.contains(currentWord.word.trimmingCharacters(in: .whitespaces)) {
```

Comment:
Getting a crash at `mergedAlignment[mergedAlignment.count - 1]` when `i == 1` and `mergedAlignment` is empty
```suggestion
            if i > 1, currentWord.word.starts(with: " "), prepended.contains(currentWord.word.trimmingCharacters(in: .whitespaces)) {
```

Reply:
actually it seems like the right fix might be to change `for i in 1..<` to `for i in 0..<`, otherwise it drops the first token.

Reply:
Based on your video I think the merge strategy needs a little re-work, notice the contractions being attached to the next words and so forth. I'll include this for the time being but expect it to be much improved shortly.